R/MsBackendMemory.R

```diff
@@ -0,0 +1,657 @@
+#' @include hidden_aliases.R
+NULL
+
+#' @title Improved in-memory MS data backend
+#'
+#' @name MsBackendMemory
+#'
+#' @author Johannes Rainer, Sebastian Gibb, Laurent Gatto
+#'
+#' @noRd
+#'
+#' @exportClass MsBackendMemory
+NULL
+
+#' - spectraData: `data.frame` with arbitrary spectra annotations.
+#' - peaksData: `list` of `numeric` matrices with m/z and intensity values.
+#' - peaksDataFrame: `list` of `data.frame`s with potentially arbitrary peak
+#'     annotations.
+#'
+#' @noRd
+setClass("MsBackendMemory",
+         contains = "MsBackend",
+         slots = c(spectraData = "data.frame",
+                   peaksData = "list",
+                   peaksDataFrame = "list"),
+         prototype = prototype(spectraData = data.frame(),
+                               peaksData = list(),
+                               peaksDataFrame = list(),
+                               readonly = FALSE,
+                               version = "0.1"))
+
+setValidity("MsBackendMemory", function(object) {
+    msg <- .valid_spectra_data_required_columns(object@spectraData)
+    if (length(msg))
+        return(msg)
+    msg <-
+        .valid_column_datatype(object@spectraData, .SPECTRA_DATA_COLUMNS)
+    if (length(object@peaksDataFrame) && length(object@peaksData) !=
+        length(object@peaksDataFrame))
+        msg <- c(msg, "peaksData and peaksDataFrame have different length")
+    if (is.null(msg)) TRUE
+    else msg
+})
+
+#' @rdname hidden_aliases
+setMethod("show", "MsBackendMemory", function(object) {
+    spd <- spectraData(object, c("msLevel", "rtime", "scanIndex"))
+    cat(class(object), "with", nrow(spd), "spectra\n")
+    if (nrow(spd)) {
+        txt <- capture.output(print(spd))
+        cat(txt[-1], sep = "\n")
+        sp_cols <- spectraVariables(object)
+        cat(" ...", length(sp_cols) - 3, "more variables/columns.\n")
+    }
+})
+
+#' @importMethodsFrom S4Vectors $ $<-
+#'
+#' @importClassesFrom IRanges NumericList
+#'
+#' @importFrom IRanges NumericList
+#'
+#' @rdname MsBackend
+setMethod("backendInitialize", signature = "MsBackendMemory",
+          function(object, data, ...) {
+              if (missing(data)) data <- data.frame()
+              if (is(data, "DataFrame"))
+                  data <- as(data, "data.frame")
```

Comment:
I would avoid that: `setBackend` calls also this function, and will submit all the data from the *old* backend as a `DataFrame` to the new backend. Thus, the user would always get this warning whenever he/she changes to an `MsBackendMemory` backend.

But like I mentioned to Laurent below (above?), I think we should add a check to the `<-,MsBackendMemory` to test whether `value` is an `S4` object - because storing them in the (internal) `data.frame` could eventually not work.

Reply:
Update: seems we can even have `S4` objects in columns of a `data.frame`. Tested with `List`, `NumericList`.